subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformWithDependenciesIntegrationTest.groovy

```diff
@@ -325,4 +329,220 @@ project(':app') {
+                }
+            }
+        """
+
+        when:
+        run "resolveSize", "resolveInter", "resolveFinal"
+        then:
+        output.count("Transforming") == 20
+        [
+            'common.jar': [],
+            'hamcrest-core-1.3.jar': [],
+            'junit-4.11.jar': ['hamcrest-core-1.3.jar']
+        ].each { artifact, dependencies ->
+            assert output.contains("Single step transform received dependencies files [${dependencies.join(", ")}] for processing ${artifact}")
+            assert output.contains("Transform step 1 received dependencies files [${dependencies.join(", ")}] for processing ${artifact}")
+            assert output.contains("Transform step 2 received dependencies files [${dependencies.collect { it + ".txt" }.join(", ")}] for processing ${artifact}.txt")
+        }
+        [
+            'slf4j-api-1.7.25.jar': [],
+            'lib.jar': ['slf4j-api-1.7.25.jar', 'common.jar']
+        ].each { artifact, dependencies ->
+            assert output.contains("Single step transform received dependencies files [${dependencies.join(", ")}] for processing ${artifact}")
+            assert output.contains("Transform step 1 received dependencies files [${dependencies.join(", ")}] for processing ${artifact}")
+        }
+        [
+            'slf4j-api-1.7.26.jar': [],
+            'lib.jar': ['slf4j-api-1.7.26.jar', 'common.jar']
+        ].each { artifact, dependencies ->
+            assert output.contains("Single step transform received dependencies files [${dependencies.join(", ")}] for processing ${artifact}")
+            assert output.contains("Transform step 1 received dependencies files [${dependencies.join(", ")}] for processing ${artifact}")
+            assert output.contains("Transform step 2 received dependencies files [${dependencies.collect { it + ".txt" }.join(", ")}] for processing ${artifact}.txt")
+        }
+
+        // TODO wolfs: Looks like we execute this second step once with the first step having dependencies slf4j-api-1.7.26.jar and once with dependencies slf4j-api-1.7.25.jar
+        output.count("Transform step 2 received dependencies files [slf4j-api-1.7.26.jar.txt, common.jar.txt] for processing lib.jar.txt") == 2
+
+        def outputLines = output.readLines()
+        def app1Resolve = outputLines.indexOf("> Task :app:resolveSize")
+        def app2Resolve = outputLines.indexOf("> Task :app2:resolveSize")
+        def lib1Transform = outputLines.indexOf("Single step transform received dependencies files [slf4j-api-1.7.25.jar, common.jar] for processing lib.jar")
+        def lib2Transform = outputLines.indexOf("Single step transform received dependencies files [slf4j-api-1.7.26.jar, common.jar] for processing lib.jar")
+        ![app1Resolve, app2Resolve, lib1Transform, lib2Transform].contains(-1)
+        assert lib1Transform < app1Resolve // scheduled
+        assert lib2Transform > app2Resolve // immediate, TODO wolfs: should be scheduled as well
```

Comment:
A bit more explanation about what's happening here, and what should be happening here, would be nice.